File: src/lib/keywords.ts
const STOPWORDS = new Set([
  "a",
  "about",
  "above",
  "after",
  "again",
  "all",
  "also",
  "am",
  "an",
  "and",
  "any",
  "are",
  "as",
  "at",
  "be",
  "because",
  "been",
  "before",
  "being",
  "below",
  "between",
  "both",
  "but",
  "by",
  "can",
  "candidate",
  "company",
  "could",
  "day",
  "did",
  "do",
  "does",
  "doing",
  "down",
  "during",
  "each",
  "few",
  "for",
  "from",
  "further",
  "had",
  "has",
  "have",
  "having",
  "he",
  "her",
  "here",
  "hers",
  "him",
  "his",
  "how",
  "i",
  "if",
  "in",
  "into",
  "is",
  "it",
  "its",
  "job",
  "just",
  "more",
  "most",
  "must",
  "my",
  "no",
  "nor",
  "not",
  "of",
  "off",
  "on",
  "once",
  "only",
  "or",
  "other",
  "our",
  "out",
  "over",
  "own",
  "per",
  "role",
  "same",
  "she",
  "should",
  "so",
  "some",
  "such",
  "than",
  "that",
  "the",
  "their",
  "theirs",
  "them",
  "then",
  "there",
  "these",
  "they",
  "this",
  "those",
  "through",
  "to",
  "too",
  "under",
  "until",
  "up",
  "very",
  "was",
  "we",
  "were",
  "what",
  "when",
  "where",
  "which",
  "while",
  "who",
  "whom",
  "why",
  "will",
  "with",
  "work",
  "would",
  "you",
  "your"
]);

const SKILL_TERMS = [
  "accessibility",
  "agile",
  "airflow",
  "analytics",
  "api",
  "aws",
  "azure",
  "backend",
  "bigquery",
  "ci/cd",
  "cloud",
  "collaboration",
  "communication",
  "computer vision",
  "css",
  "customer success",
  "data analysis",
  "data engineering",
  "data pipelines",
  "data science",
  "devops",
  "docker",
  "etl",
  "fastapi",
  "figma",
  "firebase",
  "frontend",
  "gcp",
  "graphql",
  "html",
  "java",
  "javascript",
  "kubernetes",
  "langchain",
  "leadership",
  "llm",
  "machine learning",
  "microservices",
  "mongodb",
  "next.js",
  "node.js",
  "openai",
  "postgresql",
  "product design",
  "python",
  "react",
  "redis",
  "rest",
  "salesforce",
  "sql",
  "stakeholder management",
  "supabase",
  "tailwind",
  "terraform",
  "typescript",
  "ui/ux",
  "vercel"
];

export function normalizeText(value: string) {
  return value
    .toLowerCase()
    .replace(/[\u2018\u2019]/g, "'")
    .replace(/[\u201c\u201d]/g, '"')
    .replace(/[^a-z0-9+#./\s-]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

export function tokenize(value: string) {
  return normalizeText(value)
    .split(/\s+/)
    .map((token) => token.trim())
    .filter((token) => token.length > 2 && !STOPWORDS.has(token));
}

export function extractKeywords(value: string, limit = 35) {
  const normalized = normalizeText(value);
  const tokens = tokenize(normalized);
  const counts = new Map<string, number>();

  for (const token of tokens) {
    counts.set(token, (counts.get(token) ?? 0) + 1);
  }

  const skillHits = SKILL_TERMS.filter((term) => normalized.includes(term));
  for (const skill of skillHits) {
    counts.set(skill, (counts.get(skill) ?? 0) + 5);
  }

  return Array.from(counts.entries())
    .sort((a, b) => b[1] - a[1] || b[0].length - a[0].length)
    .slice(0, limit)
    .map(([term]) => term);
}

export function inferJobDetails(jd: string) {
  const lines = jd
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .slice(0, 20);

  let jobTitle = "";
  let companyName = "";

  for (const line of lines) {
    const titleMatch = line.match(/(?:job\s*title|title|role)\s*[:\-]\s*(.+)$/i);
    if (titleMatch?.[1] && !jobTitle) {
      jobTitle = titleMatch[1].trim();
    }

    const companyMatch = line.match(/(?:company|organization|employer)\s*[:\-]\s*(.+)$/i);
    if (companyMatch?.[1] && !companyName) {
      companyName = companyMatch[1].trim();
    }
  }

  if (!jobTitle && lines[0]) {
    jobTitle = lines[0].replace(/^we are hiring\s*/i, "").slice(0, 80);
  }

  return {
    jobTitle: jobTitle || "Target Role",
    companyName: companyName || "Target Company"
  };
}

export function formatBytes(size: number) {
  if (size < 1024) {
    return `${size} B`;
  }
  if (size < 1024 * 1024) {
    return `${(size / 1024).toFixed(1)} KB`;
  }
  return `${(size / (1024 * 1024)).toFixed(1)} MB`;
}

File: app/keywords.py
from __future__ import annotations

import re
from collections import Counter

STOPWORDS = {
    "a",
    "about",
    "above",
    "after",
    "all",
    "also",
    "an",
    "and",
    "any",
    "are",
    "as",
    "at",
    "be",
    "been",
    "being",
    "by",
    "can",
    "candidate",
    "company",
    "do",
    "does",
    "for",
    "from",
    "has",
    "have",
    "in",
    "into",
    "is",
    "it",
    "job",
    "more",
    "most",
    "must",
    "of",
    "on",
    "or",
    "our",
    "role",
    "that",
    "the",
    "their",
    "this",
    "to",
    "we",
    "with",
    "work",
    "you",
    "your",
}

SKILL_TERMS = [
    "accessibility",
    "agile",
    "airflow",
    "analytics",
    "api",
    "aws",
    "azure",
    "backend",
    "cloud",
    "communication",
    "css",
    "data analysis",
    "data engineering",
    "devops",
    "docker",
    "fastapi",
    "frontend",
    "gcp",
    "graphql",
    "html",
    "javascript",
    "kubernetes",
    "llm",
    "machine learning",
    "next.js",
    "node.js",
    "openai",
    "postgresql",
    "product design",
    "python",
    "react",
    "sql",
    "stakeholder management",
    "typescript",
    "ui/ux",
]


def normalize_text(value: str) -> str:
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9+#./\s-]", " ", value.lower())).strip()


def tokenize(value: str) -> list[str]:
    return [token for token in normalize_text(value).split() if len(token) > 2 and token not in STOPWORDS]


def extract_keywords(value: str, limit: int = 35) -> list[str]:
    normalized = normalize_text(value)
    counts = Counter(tokenize(normalized))

    for term in SKILL_TERMS:
        if term in normalized:
            counts[term] += 5

    return [term for term, _ in counts.most_common(limit)]


def infer_job_details(jd: str) -> dict[str, str]:
    lines = [line.strip() for line in jd.splitlines() if line.strip()][:20]
    job_title = ""
    company_name = ""

    for line in lines:
        title_match = re.search(r"(?:job\s*title|title|role)\s*[:\-]\s*(.+)$", line, flags=re.I)
        company_match = re.search(r"(?:company|organization|employer)\s*[:\-]\s*(.+)$", line, flags=re.I)
        if title_match and not job_title:
            job_title = title_match.group(1).strip()
        if company_match and not company_name:
            company_name = company_match.group(1).strip()

    if not job_title and lines:
        job_title = re.sub(r"^we are hiring\s*", "", lines[0], flags=re.I)[:80]

    return {
        "job_title": job_title or "Target Role",
        "company_name": company_name or "Target Company",
    }


def format_bytes(size: int) -> str:
    if size < 1024:
        return f"{size} B"
    if size < 1024 * 1024:
        return f"{size / 1024:.1f} KB"
    return f"{size / (1024 * 1024):.1f} MB"
